src/github.com/matrix-org/dendrite/clientapi/auth/storage/accounts/filter_table.go

```diff
@@ -71,20 +72,29 @@ func (s *filterStatements) prepare(db *sql.DB) (err error) {
 
 func (s *filterStatements) selectFilter(
 	ctx context.Context, localpart string, filterID string,
-) (filter []byte, err error) {
-	err = s.selectFilterStmt.QueryRowContext(ctx, localpart, filterID).Scan(&filter)
-	return
+) (*gomatrix.Filter, error) {
+	var filterData []byte
+	err := s.selectFilterStmt.QueryRowContext(ctx, localpart, filterID).Scan(&filterData)
+	if err != nil {
+		return nil, err
+	}
+
+	var filter gomatrix.Filter
+	if err = json.Unmarshal(filterData, &filter); err != nil {
+		return nil, err
+	}
+	return &filter, err
 }
 
 func (s *filterStatements) insertFilter(
-	ctx context.Context, filter []byte, localpart string,
+	ctx context.Context, filter *gomatrix.Filter, localpart string,
 ) (filterID string, err error) {
 	var existingFilterID string
 
 	// This can result in a race condition when two clients try to insert the
 	// same filter and localpart at the same time, however this is not a
 	// problem as both calls will result in the same filterID
-	filterJSON, err := gomatrixserverlib.CanonicalJSON(filter)
+	filterJSON, err := json.Marshal(filter)
```

Comment:
I suspect there was a reason it was encoding as CanonicalJSON.

Reply:
gomatrixserverlib.CanonicalJSON removes whitespaces & sorts elements, including array content (which is a bit weird IMO since it will cause issues if the array elements need to be in a specific order, but it's not the case for filters) without building a JSON object.

json.Marshall already produces whitespace-less json, but the ordering (I suppose) depends on the hash value of the key, rather than alphabetical order.

Thanks for pointing this, I just found out that in the insert query, there was a check to prevent inserting multiple times the same filter, by comparing contents. I will use `gomatrixserverlib.CanonicalJSON(json.Marshal(filter))` when splitting this PR